### ISSUE TITLE
Document the meaning of --set-target=default vs reject

### ISSUE THREAD

**Issue by erig0** [firewalld/firewalld#590]
"default" is almost, but not quite the same as "reject". The major difference is in the FORWARD chain in which packets can fall to the "OUT" zones (see issue #177 and item 2 below).

1) source-based zone fall-though/drift to interface-based zones

  - This is known as "zone drifting" and was recently addressed in #441.
  - If a packet ingress (INPUT) a source-based zone, it may still enter
    an interface-based zone (including the default zone)

2) For FORWARD traffic, packets that ingress a zone with a "default"
   target will be allowed based on the target of the egress zone.

   A forwarded packet that ingresses zoneA and egresses zoneB.

     - if zoneA target is "ACCEPT", "DROP", or "%%REJECT%%" the packet
       accepted, dropped, or rejected repectively.
     - if zoneA target is "default", then the packet is
       accepted/dropped/rejected based on zoneB's target.

3) with "default" ICMP is explicitly accepted at the end of the zone's rule set, but other packets will fall through to the top-level reject

**Comment by toreanderson** [comment]
Possible dup of #210?

For what it's worth, I headed down this rabbit hole as `DROP` and `%%REJECT%%` appeared to be totally unusable because they drop inbound IPv6 Neighbour Discovery packets. ND is a fundamental part of the IPv6 protocol, so the network attachment is as a result completely broken. (The IPv4 equivalent would be to drop ARP packets.) The `default` target, on the other hand, seemed to work fine - but I could not find any documentation that explained if there are any other differences between `default` and `DROP`/`%%REJECT%%`.

**Comment by erig0** [comment]
@toreanderson, I added #3 to above which cover the ICMP topic you mentioned.

**Comment by erig0** [comment]
Closing #210 in favor of this report as this one has more useful information.

**Comment by toreanderson** [comment]
> @toreanderson, I added item number 3 above which covers the ICMP topic you mentioned.
> 
> As you have found, if the zone uses DROP or REJECT, then you must explicitly allow ICMP. e.g.
> 
> ```
> # firewalld-cmd --zone <zone> --add-protocol icmp
> ```

Thanks for pointing that out. That said, I feel that IPv6 ND ought to be an exception to this, and allowed by default even in `DROP` and `REJECT` zones as well.

My reasoning is is that ND is a fundamental and essential part of the IPv6 protocol, in exactly the same way that ARP is an essential part of IPv4. It appears that`firewalld` allows ARP by default in  `DROP` and `REJECT` zones, so in order to attain parity between IPv4 and IPv6, you would need to allow ND by default as well (or deny ARP by default).

Furthermore, I would understand that the zone target is meant to control what happens to inbound traffic. However, when you break ND, you also break outbound traffic, because the node running `firewalld` cannot look up the MAC addresses of its link neighbours - including the default gateway. Thus all outbound traffic ends up being blackholed as well.

Honestly, I cannot think of a realistic scenario where you'd actually want to break ND (or ARP) by default.

**Comment by erig0** [comment]
On Wed, Mar 25, 2020 at 10:49:31AM -0700, Tore Anderson wrote:
> > @toreanderson, I added item number 3 above which covers the ICMP topic you mentioned.
> > 
> > As you have found, if the zone uses DROP or REJECT, then you must explicitly allow ICMP. e.g.
> > 
> > ```
> > # firewalld-cmd --zone <zone> --add-protocol icmp
> > ```
> 
> Thanks for pointing that out. That said, I feel that IPv6 ND ought to
> be an exception to this, and allowed by default even in `DROP` and
> `REJECT` zones as well.

Sounds like you want the behavior of `--set-target=default`.

> My reasoning is is that ND is a fundamental and essential part of the
> IPv6 protocol, in exactly the same way that ARP is an essential part
> of IPv4. It appears that`firewalld` allows ARP by default in  `DROP`
> and `REJECT` zones, so in order to attain parity between IPv4 and
> IPv6, you would need to allow ND by default as well (or deny ARP by
> default).

ARP is not explicitly allowed by firewalld. ARP is allowed by the
kernel. Why ARP and ND behave differently? ARP is _not_ an IPv4 packet
and ND _is_ an IPv6 packet. Firewalld does _not_ do any ARP filtering
(arptables, nftables arp family).

I disagree with you. Since ND _is_ an IPv6 packet and the user is using
a drop zone, then we should drop the packets. User is responsible for
allowing the necessary traffic. We shouldn't guess what they want. We
have more friendly zones for that reason, i.e. public.

I acknowledge the incongruence between IPv4 and IPv6, but I still think
it's the correct behavior.

> Furthermore, I would understand that the zone target is meant to
> control what happens to inbound traffic. However, when you break ND,
> you also break outbound traffic, because the node running `firewalld`
> cannot look up the MAC addresses of its link neighbours - including
> the default gateway. Thus all outbound traffic ends up being
> blackholed as well.

It would work just fine if you manually added neighbors with iproute.

> Honestly, I cannot think of a realistic scenario where you'd actually
> want to break ND (or ARP) by default.

It's not broken by default. The user has to take action to break it.
They must assign an interface to the drop/reject zones.

Some people may want to block ND if:
    - they use static addressing
    - their network is IPv4 only

**Comment by toreanderson** [comment]
> Sounds like you want the behavior of `--set-target=default`.

I do not, actually, because of your item 2 / issue #177.

> ARP is not explicitly allowed by firewalld. ARP is allowed by the kernel.

The exact same thing is true for ND in the context of drop/reject zones.

> ARP is _not_ an IPv4 packet and ND _is_ an IPv6 packet.

True, but ARP *is* part of the IPv4 protocol suite.

> Firewalld does _not_ do any ARP filtering (arptables, nftables arp family).

My point exactly. In order to get consistent behaviour between IPv4 and IPv6, firewalld *should* have performed ARP filtering by default in the drop/reject zones *-OR-* allowed ND by default in such zones.

Filtering just one of them but not the other makes no sense whatsoever.

> Since ND _is_ an IPv6 packet and the user is using a drop zone, then we should drop the packets. User is responsible for allowing the necessary traffic. We shouldn't guess what they want. We have more friendly zones for that reason, i.e. public.

«Since ARP _is_ an IPv4 protocol family packet and the user is using a drop zone, then we should drop the packets. User is responsible for allowing the necessary traffic. We shouldn't guess what they want. We have more friendly zones for that reason, i.e. public.»

> It would work just fine if you manually added neighbors with iproute.

On *all* nodes in the network, not just the one(s) running firewalld. (Nobody runs their network like this.)

**Comment by erig0** [comment]
On Wed, Mar 25, 2020 at 02:14:58PM -0700, Tore Anderson wrote:
> > ARP is _not_ an IPv4 packet and ND _is_ an IPv6 packet.
> 
> True, but ARP *is* part of the IPv4 protocol suite.

By that logic we should allow IGMP by default as well.

> > Firewalld does _not_ do any ARP filtering (arptables, nftables arp family).
> 
> My point exactly. In order to get consistent behaviour between IPv4
> and IPv6, firewalld *should* have performed ARP filtering by default
> in the drop/reject zones *-OR-* allowed ND by default in such zones.
> 
> Filtering just one of them but not the other makes no sense
> whatsoever.

Feel free to file a separate issue for ND in drop/reject zones so we can
continue this topic and not muddy this issue further.

I see both sides of the argument, but I tend to lean towards not
changing existing behavior.

> > Since ND _is_ an IPv6 packet and the user is using a drop zone, then
> > we should drop the packets. User is responsible for allowing the
> > necessary traffic. We shouldn't guess what they want. We have more
> > friendly zones for that reason, i.e. public.
> 
> «Since ARP _is_ an IPv4 protocol family packet and the user is using a
> drop zone, then we should drop the packets. User is responsible for
> allowing the necessary traffic. We shouldn't guess what they want. We
> have more friendly zones for that reason, i.e. public.»

Maybe true, but we can't change that behavior without breaking users.

**Comment by toreanderson** [comment]
> By that logic we should allow IGMP by default as well.

If MLD is allowed by default in drop/reject zones, then it would indeed make perfect sense to allow IGMP by default as well. If only one of them was allowed and the other blocked, that would be inconsistent.

As I understand it, however, both IGMP and MLD are blocked by default in drop/reject zones, which is perfectly sensible. After all, unlike ARP and ND, IGMP and MLD are not essential for the correct functioning of a network. Most networks operate fine without IGMP and MLD.

> Feel free to file a separate issue for ND in drop/reject zones so we can continue this topic and not muddy this issue further. I see both sides of the argument, but I tend to lean towards not changing existing behavior.

OK - I will do that.

> > «Since ARP _is_ an IPv4 protocol family packet and the user is using a drop zone, then we should drop the packets. User is responsible for allowing the necessary traffic. We shouldn't guess what they want. We have more friendly zones for that reason, i.e. public.»
> >
> Maybe true, but we can't change that behavior without breaking users.

But that's kind of the argument you're making - users using drop/reject zones apparently *want* their (IPv6) networks to be totally broken by default, not able to send traffic in any direction. Applying that logic, the fact that their IPv4 networks keeps working when using drop/reject zones is a bug. :slightly_smiling_face:

**Comment by erig0** [comment]
On Thu, Mar 26, 2020 at 10:36:36AM -0700, Tore Anderson wrote:
> > > «Since ARP _is_ an IPv4 protocol family packet and the user is using a drop zone, then we should drop the packets. User is responsible for allowing the necessary traffic. We shouldn't guess what they want. We have more friendly zones for that reason, i.e. public.»
> > >
> > Maybe true, but we can't change that behavior without breaking users.
> 
> But that's kind of the argument you're making - users using drop/reject zones apparently *want* their (IPv6) networks to be totally broken by default, not able to send traffic in any direction. Applying that logic, the fact that their IPv4 networks keeps working when using drop/reject zones is a bug. :slightly_smiling_face: 

I meant changing firewalld's behavior may surprise users. i.e. "break"
as in "breaking change".

**Comment by erig0** [comment]
Here is the proposed new entry for the man page. I would appreciate feedback. @toreanderson
```
--permanent [--zone=zone] --set-target=target
    Set the target of a permanent zone.  target is one of: default,
    ACCEPT, DROP, REJECT

    default is similar to REJECT, but has special meaning in the
    following scenarios:

     1. ICMP explicitly allowed

        At the end of the zone's ruleset ICMP packets are explicitly
        allowed.

     2. forwarded packets follow the target of the egress zone

        In the case of forwarded packets if the ingress zone uses
        default then whether or not the packet will be allowed is
        determined by the egress zone.

        For a forwarded packet that ingresses zoneA and egresses zoneB:

        ·   if zoneA's target is ACCEPT, DROP, or REJECT then the
            packet is accepted, dropped, or rejected respectively.

        ·   if zoneA's target is default, then the packet is accepted,
            dropped, or rejected based on zoneB's target.

     3. Zone drifting from source-based zone to interface-based zone

        This only applies if AllowZoneDrifting is enabled. See
        firewalld.conf(5).

        If a packet ingresses a source-based zone, it may still enter
        an interface-based zone (including the default zone).
```

**Comment by toreanderson** [comment]
I found this very clear. :+1: However, I have one question, regarding item 2: What happens to forwarded packets if both zoneA and zoneB have the `default` target?

**Comment by erig0** [comment]
@toreanderson, I edited my comment. I added the sentence "If zoneB's target is also default, then the packet will be rejected by firewalld's catchall reject." Does this make it clear?

**Comment by toreanderson** [comment]
Yes, crystal clear! :+1:

**Comment by snovotill** [comment]
Requesting that `target=default` be included in the document `man firewalld.xml` sections cited below. I had a particularly bad time figuring this out until I found it nicely documented in man `firewall-cmd`. The current description mentions "default" in a different context, which is what threw me.

`<zone [version="versionstring"] [target="ACCEPT|%%REJECT%%|DROP"] [ingress-priority="priority"] [egress-priority="priority"]>`
and
```
target="ACCEPT|%%REJECT%%|DROP"
           Can be used to accept, reject or drop every packet that doesn't match any rule (port, service, etc.). The ACCEPT target is used in trusted zone to accept
           every packet not matching any rule. The %%REJECT%% target is used in block zone to reject (with default firewalld reject type) every packet not matching
           any rule. The DROP target is used in drop zone to drop every packet not matching any rule. If the target is not specified, every packet not matching any
           rule will be rejected.
```
